### PR TITLE
(WIP) Merge ethereum upstream

### PR DESCRIPTION
**Pull request by screwyprof**

### CODE REVIEW

File: consensus/ethash/sealer.go [hunk 1]
```diff
@@ -186,186 +186,255 @@ search:
...
+	s.works[hash] = block
+}
+
+// notifyWork notifies all the specified mining endpoints of the availability of
+// new work to be processed.
+func (s *remoteSealer) notifyWork() {
+	work := s.currentWork
+	blob, _ := json.Marshal(work)
+	s.reqWG.Add(len(s.notifyURLs))
+	for _, url := range s.notifyURLs {
```
Comment (JekaMas) [inline consensus/ethash/sealer.go]: Possible leak

File: consensus/ethash/sealer.go [hunk 2]
```diff
@@ -186,186 +186,255 @@ search:
...
+		rates:        make(map[common.Hash]hashrate),
+		workCh:       make(chan *sealTask),
+		fetchWorkCh:  make(chan *sealWork),
+		submitWorkCh: make(chan *mineResult),
+		fetchRateCh:  make(chan chan uint64),
+		submitRateCh: make(chan *hashrate),
+		requestExit:  make(chan struct{}),
+		exitCh:       make(chan struct{}),
 	}
+	go s.loop()
```
Comment (JekaMas) [inline consensus/ethash/sealer.go]: A possible leak

File: core/blockchain_test.go [hunk 3]
```diff
@@ -1060,40 +1073,31 @@ func TestLogRebirth(t *testing.T) {
...
-	case <-time.NewTimer(1 * time.Second).C:
-		t.Fatal("Timeout. There is no RemovedLogsEvent has been sent.")
+	if !<-removeLogCh {
+		t.Fatal("failed to receive removed log event")
 	}
 
 	newBlocks, _ := GenerateChain(params.TestChainConfig, chain[len(chain)-1], ethash.NewFaker(), db, 1, func(i int, gen *BlockGen) {})
-	go listenNewLog(logsCh, 1)
+	go validateLogEvent(logsCh, newLogCh, 1)
+	go validateLogEvent(rmLogsCh, removeLogCh, 1)
```
Comment (JekaMas) [inline core/blockchain_test.go]: Possible leaks

File: core/blockchain_test.go [hunk 4]
```diff
@@ -1038,12 +1051,12 @@ func TestLogRebirth(t *testing.T) {
 	})
 
 	// Spawn a goroutine to receive log events
-	go listenNewLog(logsCh, 1)
+	go validateLogEvent(logsCh, newLogCh, 1)
```
Comment (JekaMas) [inline core/blockchain_test.go]: A possible leak

**Comment by screwyprof** [comment]
> Why did you remove all this tests?
I didn't the filed was renamed in upstreamed and moved to `urlv4_test.go`

**Comment by screwyprof** [comment]
Fixed by #386

**Pull request closed without merging**